Spring controller add smooth mesh and static camera height tracking

dzosz opened the pull request:
Fixes #708

SpringController - track airmesh height instead of ground  for smooth horizontal camera movements
* add bool option CamSpringTrackMapHeight (default=1) to disable tracking the ground height entirely. with this option disabled camera can move up if it's below the ground of the hill but will never go back down by itself

dzosz commented:
Is smooth mesh always available? Any additional checking required?

badosu commented on rts/Game/Camera/SpringController.cpp:
```diff
@@ -24,7 +25,16 @@ CONFIG(bool,  CamSpringZoomOutFromMousePos).defaultValue(false);
 CONFIG(bool,  CamSpringEdgeRotate).defaultValue(false).description("Rotate camera when cursor touches screen borders.");
 CONFIG(float, CamSpringFastScaleMouseMove).defaultValue(3.0f / 10.0f).description("Scaling for CameraMoveFastMult in spring camera mode while moving mouse.");
 CONFIG(float, CamSpringFastScaleMousewheelMove).defaultValue(2.0f / 10.0f).description("Scaling for CameraMoveFastMult in spring camera mode while scrolling with mouse.");
+CONFIG(bool, CamSpringTrackMapHeight).defaultValue(true).description("Camera height is influenced by terrain height");
```
Perhaps I'd rename it to something that indicates the height tracking and make it an int, e.g. if in the future tracking constant height instead of airmesh if desired etc.

lhog commented:
> Is smooth mesh always available? Any additional checking required?

@marcushutchings to comment. IIRC it's not always used, but I'm not sure it means it doesn't exist.

dzosz commented:
@badosu @PtaQQ do we need the legacy camera behavior preserved? E.g. with option CamSpringTrackMapHeight=2

marcushutchings commented:
> Is smooth mesh always available? Any additional checking required?

It can be switched off yes. `bool system.enableSmoothMesh = false` see <https://github.com/beyond-all-reason/spring/wiki/Mod-Rules>

You can query that parameter from Sim/Misc/ModInfo.h

marcushutchings commented:
> @badosu @PtaQQ do we need the legacy camera behavior preserved? E.g. with option CamSpringTrackMapHeight=2

Generally, yes. We have to consult the game devs as a whole before we remove features. Removal of features hould be handled in a spearate Issue/PR.

badosu commented:
I didn't notice default, until discussion with other consumers the default should be set to 105 expected behavior unless it's a bugfix.

dzosz commented:
Changes:
* default behavior not changed
* config value CamSpringTrackMapHeightMode has now three options: 0 for static camera height, 1 (default) for tracking terrain height, 2 for tracking airmesh
* cleaned up code so the change is more incremental
* added bound checks to fix quirky behavior close to the map boundaries

sprunk commented on rts/Game/Camera/SpringController.cpp:
```diff
@@ -54,13 +63,44 @@ void CSpringController::ConfigUpdate()
+		smoothGround.GetHeight(pos.x, pos.z) - smoothGround.GetHeight(prevPos.x, prevPos.z) :
+		0.0f;
+
+	if (trackMapHeight == HeightTracking::Smooth || trackMapHeight == HeightTracking::Disabled) {
```
Looks suspicious that not only is `HeightTracking::Disabled` handled in some way but also it's in the function supposedly dedicated for `HeightTracking::Smooth`.

sprunk commented on rts/Game/Camera/SpringController.cpp:
```diff
@@ -24,7 +33,7 @@ CONFIG(bool,  CamSpringZoomOutFromMousePos).defaultValue(false);
 CONFIG(float, CamSpringFastScaleMouseMove).defaultValue(3.0f / 10.0f).description("Scaling for CameraMoveFastMult in spring camera mode while moving mouse.");
 CONFIG(float, CamSpringFastScaleMousewheelMove).defaultValue(2.0f / 10.0f).description("Scaling for CameraMoveFastMult in spring camera mode while scrolling with mouse.");
-
+CONFIG(int,   CamSpringTrackMapHeightMode).defaultValue(HeightTracking::Terrain).description("Camera height is influenced by terrain height. 0=Static 1=Terrain 2=Smooth");
```
"Smooth" is a fairly generic term, perhaps would be best to be precise
```suggestion
CONFIG(int,   CamSpringTrackMapHeightMode).defaultValue(HeightTracking::Terrain).description("Camera height is influenced by terrain height. 0=Static 1=Terrain 2=Smoothmesh");
```

sprunk commented on rts/Game/Camera/SpringController.cpp:
```diff
@@ -54,13 +63,44 @@ void CSpringController::ConfigUpdate()
+		return;
+	}
+
+	const float camHeightDiff = (trackMapHeight == HeightTracking::Smooth) ?
+		smoothGround.GetHeight(pos.x, pos.z) - smoothGround.GetHeight(prevPos.x, prevPos.z) :
+		0.0f;
```
This value doesn't seem used outside the `if` scope.

sprunk commented on rts/Game/Camera/SpringController.cpp:
```diff
@@ -75,9 +115,14 @@ void CSpringController::KeyMove(float3 move)
 	const float3 flatForward = (dir * XZVector).ANormalize();
 	pos += (camera->GetRight() * move.x + flatForward * move.y) * pixelSize * 2.0f * scrollSpeed;
+
+	SmoothCamHeight(prevPos);
```
How about, instead of the mode check inside `SmoothCamHeight()` in line 88, something like this?
```cpp
	switch (trackMapHeight) {

	case HeightTracking::Disabled:
		/* also calls SmoothCamHeight, because
		 * of reason <FIXME: fill in reason> */
		[[fallthrough]];

	case HeightTracking::Smooth:
		SmoothCamHeight(prevPos);
		break;

	case HeightTracking::Terrain:
		/* tracking the terrain is handled at <FIXME: where?>,
		 * so we don't do anything here */
		break;
	}
```

dzosz replied:
that's much better

dzosz commented:
@sprunk thanks for review. your suggestions are implemented

lhog commented:
Does it make sense to add a runtime switch for that `/camtrackmapheight` perhaps?
I know you can `/set CamSpringTrackMapHeightMode <mode>`, but it's a bit more involved.

sprunk commented on rts/Game/Camera/SpringController.cpp:
```diff
@@ -54,13 +63,41 @@ void CSpringController::ConfigUpdate()
+	camPos.y = std::max(camPos.y, CGround::GetHeightReal(camPos.x, camPos.z, false) + 5.0f);
+
+	// raycast to ground to simulate camera movement and find new point of focus
+	const float distToGround = CGround::LineGroundCol(camPos, camPos + dir * 150000.0f, false);
```
consider some explanation for the magic 150k constant

lhog replied:
I think it's ok as is, the rest of the code using rays tracing is full of magic constants.

lhog commented on rts/Game/Camera/SpringController.cpp:
```diff
@@ -54,13 +63,41 @@ void CSpringController::ConfigUpdate()
+
+	// raycast to ground to simulate camera movement and find new point of focus
+	const float distToGround = CGround::LineGroundCol(camPos, camPos + dir * 150000.0f, false);
+	// FIXME camera focus is now first ground intersection which is not what we want
```
Also @dzosz perhaps you can consider using `readMap->GetUnsyncedHeightInfo(x_hm, z_hm)`, which stores min,max and average of a height-map patch 1024x1024 size in world coordinates. When the ray tracing is complete and you have a world coordinate where the ray hits the ground you can average the average or max height of some number of nearby "patches" and use that as a height guidance. Just an idea.

dzosz replied:
Would it solve the following problem? Rotate the camera so the previous point of focus is hidden behind it. We don't know what exact (x,z) we want without raytracing through the hill. Raytracing multiple times would work but not guarantee that he last sucessful raytrace is the actual position we want (if the camera angle is very flat or map points downhill we might focus on the point.
![image](https://user-images.githubusercontent.com/13596691/232152155-f1037910-cc83-4c9f-9fcb-0f23718e2dfd.png)
where p1 is first point of focus, and p2 second point of focus (it should still be p1 though)

dzosz commented:
> Does it make sense to add a runtime switch for that `/camtrackmapheight` perhaps? I know you can `/set CamSpringTrackMapHeightMode <mode>`, but it's a bit more involved.

A runtime switch meaning an option in camera settings in BAR/BAR repo?

lhog commented:
> A runtime switch meaning an option in camera settings in BAR/BAR repo?

Nope a option you can type in the chat like `/camtrackmapheight` to turn this mode on and off

dzosz commented:
> > A runtime switch meaning an option in camera settings in BAR/BAR repo?
> 
> Nope a option you can type in the chat like `/camtrackmapheight` to turn this mode on and off

Is there any benefit to it? I expect players to enable it in GUI and developers can use /set ...

lhog commented:
Can you resolve the conflict please?
Sorry it took so long to merge.
